### PR TITLE
NVIDIA-554: DPU-host mode: use ConfigMap for OVN feature enablement instead of per-node script gating

### PR DESCRIPTION
**Pull request by tsorya**
NVIDIA-554: DPU-host mode: use ConfigMap for OVN feature enablement instead of per-node script gating

         All OVN-Kubernetes features (egress IP, egress firewall, multicast,
      multi-network, admin network policy, multi-external-gateway, etc.)
      are now enabled in DPU-host mode. The OVN controller on DPU-host
      nodes processes the configuration but does not offload egress IP
      datapath — traffic follows the regular kernel path instead.

        OVN_NODE_MODE remains used only for DPU-host structural differences:
        gateway interface, ovnkube-node-mode flag, and init-controller.

      Because these features are safe to enable cluster-wide, the per-node
      gating logic in the startup script (008-script-lib.yaml) is no
      longer needed. Feature flags are managed solely through the
      cluster-wide ConfigMap (004-config.yaml) passed to ovnkube via
      --config-file.

        Co-authored-by: Cursor <cursoragent@cursor.com>
        Made-with: Cursor

    This reverts commit 6f5697cb0b25fe1ad6cb48a9c969653475afa4b9

### CODE REVIEW

**Comment by tsorya** [comment]
oc exec -it -n openshift-ovn-kubernetes                           ovnkube-node-bpkjk -- bash [root@dhcp-8-231-235 ~]# ovn-nbctl find ACL | grep -i mcast match               : "(ip4.mcast || mldv1 || mldv2 || (ip6.dst[120..127] == 0xff && ip6.dst[116] == 1))" match               : "(ip4.mcast || mldv1 || mldv2 || (ip6.dst[120..127] == 0xff && ip6.dst[116] == 1))" match               : "outport == @a15696136357465712812 && (igmp || (ip4.src == $a14513626322132345176 && ip4.mcast))" match               : "outport == @a4743249366342378346 && (ip4.mcast || mldv1 || mldv2 || (ip6.dst[120..127] == 0xff && ip6.dst[116] == 1))" match               : "inport == @a4743249366342378346 && (ip4.mcast || mldv1 || mldv2 || (ip6.dst[120..127] == 0xff && ip6.dst[116] == 1))" match               : "inport == @a15696136357465712812 && ip4.mcast"

10.130.0.3  :   unicast, xmt/rcv/%loss = 10/10/0%, min/avg/max/std-dev = 0.290/0.404/1.026/0.222 10.130.0.3  : multicast, xmt/rcv/%loss = 10/9/9% (seq>=2 0%), min/avg/max/std-dev = 0.307/0.522/1.115/0.319 10.128.0.30 :   unicast, xmt/rcv/%loss = 10/10/0%, min/avg/max/std-dev = 0.314/0.449/0.717/0.140 10.128.0.30 : multicast, xmt/rcv/%loss = 10/9/9% (seq>=2 0%), min/avg/max/std-dev = 0.285/0.445/0.823/0.155

10.128.0.30 :   unicast, xmt/rcv/%loss = 10/10/0%, min/avg/max/std-dev = 0.307/0.488/0.991/0.218 10.128.0.30 : multicast, xmt/rcv/%loss = 10/9/9% (seq>=2 0%), min/avg/max/std-dev = 0.313/0.499/1.212/0.273 10.129.0.7  :   unicast, xmt/rcv/%loss = 10/10/0%, min/avg/max/std-dev = 0.307/0.462/1.057/0.252 10.129.0.7  : multicast, xmt/rcv/%loss = 10/9/9% (seq>=2 0%), min/avg/max/std-dev = 0.320/0.501/1.513/0.382

**Comment by yingwang-0320** [comment]
/verified by pre-merge testing.
Ran CNO and multicast cases, all passed.

**Comment by zenghui-shi** [comment]
@tsorya  hi, was there any QE testing done with this PR applied and automation/CI developed for the DPU-HOST mode? If so, would you mind pasting the testing result and CI links. I'd like to make sure we have a record for the testing on this PR and there is continuous CI job running to gate this functionality since we (core networking) currently don't have the necessary env nor CI to verify the piece of code for DPU functinoaliies.

@yingwang-0320  Sicne this change touches the enablement/disablement of several OVN-K features, could you also run testing for each of the mentioned features and see if they are functional as expected? meanwhile, like @tssurya suggested, we would also want to try enable the DPU Host mode (without testing DPU specific functionalities) and scheck if the cluster is  stable.

**Comment by zenghui-shi** [comment]
@danwinship Would you mind taking a look at this PR once the testing is done? 
This is the PR that we merged, then reverted due to multicast regression. @tsorya proposed it again and confirmed that multicast is working with the current version.

**Comment by yingwang-0320** [comment]
@tsorya we don't have any DPU-HOST mode related testing yet. Not sure what environment or card needed to enable DPU-host mode, let me confirm and update here.

**Comment by yingwang-0320** [comment]
Confirmed with @tsorya , it's OK to verify CNO on regular mode.

File: bindata/network/ovn-kubernetes/common/008-script-lib.yaml [hunk 1]
```diff
@@ -587,17 +579,10 @@ data:
         # https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5327/files
         gateway_interface="derive-from-mgmt-port"
         ovnkube_node_mode="--ovnkube-node-mode dpu-host"
-        # disable egress ip for dpu-host mode as it is not supported
-        egress_features_enable_flag=""
-
-        # disable multicast for dpu-host mode as it is not supported
-        enable_multicast_flag=""
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/common/008-script-lib.yaml:597]: The PR/commit doesn't actually explain _why_ we can get rid of all of this. Do we implement egress IPs and multicast in DPU mode now? Or does DPU mode just ignore the settings of these flags?

File: bindata/network/ovn-kubernetes/common/008-script-lib.yaml [hunk 2]
```diff
@@ -770,17 +727,9 @@ data:
         --export-ovs-metrics \
         --disable-snat-multiple-gws \
         ${export_network_flows_flags} \
-        ${multi_network_enabled_flag} \
-        ${network_segmentation_enabled_flag} \
-        ${network_connect_enabled_flag} \
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/common/008-script-lib.yaml:778]: You're removing `${network_connect_enabled_flag}` here but you still set it above.

File: bindata/network/ovn-kubernetes/managed/004-config.yaml [hunk 3]
```diff
@@ -33,14 +33,23 @@ data:
...
+    enable-egress-ip=true
+    enable-egress-firewall=true
+    enable-egress-qos=true
+    enable-egress-service=true
 {{- if .ReachabilityNodePort }}
     egressip-node-healthcheck-port={{.ReachabilityNodePort}}
 {{- end }}
-    {{- if not .OVN_MULTI_NETWORK_ENABLE }}
     enable-multi-network=true
-    {{- end }}
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/managed/004-config.yaml:41]: OK, so the handling of `enable-multi-network` is totally broken; we enable it whether `OVN_MULTI_NETWORK_ENABLE` is `true` or `false`.

It looks like when the "network segmentation" stuff was added, we made it so `DisableMultiNetwork` was ignored if the network segmentation feature gate was enabled, and then when that feature gate was removed, we ended up with a bunch of places doing "enable multi-network if `OVN_MULTI_NETWORK_ENABLE` is `true` and also enable multi-network if `OVN_MULTI_NETWORK_ENABLE` is `false`"...

You should remove all of the `DisableMultiNetwork`/`OVN_MULTI_NETWORK_ENABLE` code from the tree. (Ideally, that would be a separate commit before this one.)

File: bindata/network/ovn-kubernetes/self-hosted/ovnkube-control-plane.yaml [hunk 4]
```diff
@@ -193,23 +166,11 @@ spec:
             ${ovn_v6_transit_switch_subnet_opt} \
             ${ovn_v4_masquerade_subnet_opt} \
             ${ovn_v6_masquerade_subnet_opt} \
-            ${dns_name_resolver_enabled_flag} \
             ${persistent_ips_enabled_flag} \
-            ${multi_network_enabled_flag} \
-            ${network_segmentation_enabled_flag} \
             ${gateway_mode_flags} \
             ${network_connect_enabled_flag} \
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/self-hosted/ovnkube-control-plane.yaml]: You removed the setting of `network_connect_enabled_flag` but you forgot to remove it here.

File: bindata/network/ovn-kubernetes/managed/004-config.yaml [hunk 5]
```diff
@@ -49,7 +49,6 @@ data:
 {{- end }}
     enable-admin-network-policy=true
     enable-multi-external-gateway=true
-    enable-multicast=true
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/managed/004-config.yaml]: (I can't comment on the commit message directly; you should explain there that the issue is that there *isn't* an `enable-multicast` config file option. The flag can only be set via the CLI.)

File: pkg/network/ovn_kubernetes_test.go [hunk 6]
```diff
@@ -4460,220 +4467,6 @@ func TestRenderOVNKubernetes_AllowICMPNetworkPolicyOverride(t *testing.T) {
 	})
 }
 
-func TestOVNKubernetesControlPlaneFlags(t *testing.T) {
```
Comment (danwinship) [inline pkg/network/ovn_kubernetes_test.go:4453]: This whole commit should be squashed into the first commit. It doesn't make sense to commit a fix that breaks the unit tests first, and then have a second commit to unbreak them.

File: bindata/network/ovn-kubernetes/common/008-script-lib.yaml (hunk 1 above)
Comment (tsorya) [reply bindata/network/ovn-kubernetes/common/008-script-lib.yaml:597]: done

File: bindata/network/ovn-kubernetes/managed/004-config.yaml (hunk 3 above)
Comment (tsorya) [reply bindata/network/ovn-kubernetes/managed/004-config.yaml:41]: Done

File: bindata/network/ovn-kubernetes/common/008-script-lib.yaml [hunk 7]
```diff
@@ -727,6 +727,7 @@ data:
         --export-ovs-metrics \
         --disable-snat-multiple-gws \
         ${export_network_flows_flags} \
+        ${network_connect_enabled_flag} \
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/common/008-script-lib.yaml]: no, the bug in this file was that you were still setting `network_connect_enabled_flag` even though you were no longer using it here (and don't need to be setting it here because it's set in the config files)

**Comment by danwinship** [comment]
please squash `fix: remove broken test functions and add missing config flags in test` and `Fix stale network_connect_enabled_flag references` into the main commit (after fixing the `network_connect_enabled_flag` issue above)

File: bindata/network/ovn-kubernetes/common/008-script-lib.yaml (hunk 7 above)
Comment (tsorya) [reply bindata/network/ovn-kubernetes/common/008-script-lib.yaml]: sorry, missed it

Comment (tsorya) [reply bindata/network/ovn-kubernetes/common/008-script-lib.yaml]: Thank you !!!!!

**Comment by tsorya** [comment]
> please squash `fix: remove broken test functions and add missing config flags in test` and `Fix stale network_connect_enabled_flag references` into the main commit (after fixing the `network_connect_enabled_flag` issue above)

done

File: README.md [hunk 8]
```diff
@@ -157,18 +157,6 @@ OVNKubernetes supports the following configuration options, all of which are opt
 * `egressIPConfig`: holds the configuration for EgressIP options.
   * `reachabilityTotalTimeoutSeconds`: Set EgressIP node reachability total timeout in seconds, 0 means disable reachability check and the default is 1 second.
 
-#### DPU Host Mode Support
```
Comment (wizhaoredhat) [inline README.md:160]: Not sure why the docs are being deleted? Are they incorrect now?
We still need OVN_NODE_MODE as far as i know.

Comment (tsorya) [reply README.md:160]: This part was talking about disabled features in dpu-host mode, we are enabling them back so don't see a reason not to remove it

**Comment by tsorya** [comment]
4.22-upgrade-from-stable-4.21-e2e-azure-ovn-upgrade
	
* [sig-cli] oc adm must-gather must-gather support since and since-time flags timeout [apigroup:config.openshift.io][Timeout:20m] [Suite:openshift/conformance/parallel]

**Comment by tsorya** [comment]
•	periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview-serial-1of3: [401fc048-0800-4690-abe1-e83e251e8e7b](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-cluster-network-operator-2944-ci-4.22-e2e-aws-ovn-techpreview-serial-1of3/2049348019085643776)
Broken here is the thread around it but in this case it failed on job timeout
https://redhat-internal.slack.com/archives/CDCP2LA9L/p1776833898911429

periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview-serial-2of3
https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-cluster-network-operator-2944-ci-4.22-e2e-aws-ovn-techpreview-serial-2of3/2049348019928698880
Test succeeded failed on gathering logs, missin
[36mINFO[0m[2026-04-29T05:53:57Z] Running step e2e-aws-ovn-techpreview-serial-openshift-e2e-test. 
[36mINFO[0m[2026-04-29T07:20:15Z] Step e2e-aws-ovn-techpreview-serial-openshift-e2e-test succeeded after 1h26m17s.


	•	periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-aws-ovn-upgrade: [ec33de8c-1304-4b3f-abaa-3c10d78913c1](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/aggregator-periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-aws-ovn-upgrade/2049347947769892864)
	•	periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-azure-ovn-upgrade: [998f397c-fd2e-4652-8308-1e2da57b5540](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/aggregator-periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-azure-ovn-upgrade/2049347961195859968)
	•	periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-gcp-ovn-rt-upgrade: [72d61568-5afe-4ea5-afd0-e6f6f603b191](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/aggregator-periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-gcp-ovn-rt-upgrade/2049348074454650880)
Seems to be broken, thread around it
https://redhat-internal.slack.com/archives/C01CQA76KMX/p1777344203282859

**Comment by tsorya** [comment]
Last run
periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview-serial-3of3: c7e6565b-7fcc-4efa-ad7c-5a5f573d6100
periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview-serial-2of3: 09b35ca0-f496-4fdc-8a04-662533c7b51d
periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview-serial-1of3: e145bb5b-24e4-4040-a9f1-0666905e5df3
periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview: 1b93d83b-c9c2-4d74-9721-1ed1c0800c4a
periodic-ci-openshift-release-main-nightly-4.22-e2e-aws-ovn-serial-2of2: 18701e44-b13a-4a8e-84c8-f0c8cb805611
periodic-ci-openshift-release-main-nightly-4.22-e2e-aws-ovn-serial-1of2: c87927c2-e92e-41cb-b790-d1e0aaa4547a


`* 0001-01-01T00:00:00Z 0x : 0/118 nodes are available: 1 node(s) had untolerated taint {node.kubernetes.io/disk-pressure: }, 1 node(s) had untolerated taint {node.kubernetes.io/not-ready: }, 11 node(s) had untolerated taint {node-role.kubernetes.io/ci-prowjobs-worker: ci-prowjobs-worker}, 12 node(s) didn't match Pod's node affinity/selector, 2 node(s) had untolerated taint {node-role.kubernetes.io/ci-builds-worker: ci-builds-worker}, 24 node(s) had untolerated taint {node-role.kubernetes.io/ci-longtests-worker: ci-longtests-worker}, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }, 58 Insufficient cpu, 6 node(s) had untolerated taint {node-role.kubernetes.io/infra: }. preemption: 0/118 nodes are available: 58 No preemption victims found for incoming pod, 60 Preemption is not helpful for scheduling.
* 0001-01-01T00:00:00Z 0x : 0/118 nodes are available: 1 node(s) had untolerated taint {node.kubernetes.io/disk-pressure: }, 11 node(s) had untolerated taint {node-role.kubernetes.io/ci-prowjobs-worker: ci-prowjobs-worker}, 12 node(s) didn't match Pod's node affinity/selector, 2 node(s) didn't satisfy existing pods anti-affinity rules, 2 node(s) had untolerated taint {node-role.kubernetes.io/ci-builds-worker: ci-builds-worker}, 25 node(s) had untolerated taint {node-role.kubernetes.io/ci-longtests-worker: ci-longtests-worker}, 3 node(s) had untolerated taint {node-role.kubernetes.io/master: }, 56 Insufficient cpu, 6 node(s) had untolerated taint {node-role.kubernetes.io/infra: }. preemption: 0/118 nodes are available: 58 No preemption victims found for incoming pod, 60 Preemption is not helpful for scheduling.`

**Comment by tsorya** [comment]
latest run was better

periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-gcp-ovn-rt-upgrade: e13d2307-1204-4495-81f9-9d0f27458b4b
periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-aws-ovn-upgrade: [0c92ffd6-4fe8-40e3-88dd-dac0c30d7854](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/aggregator-periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-aws-ovn-upgrade/2050257492969000960)
periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-azure-ovn-upgrade: [4954066b-b1f1-4c72-9aee-a4a543acf56d](https://prow.ci.openshift.org/view/gs/test-platform-results/logs/aggregator-periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-azure-ovn-upgrade/2050257508966076416)
pod "network-check-target-njvzg" is using the default service account



periodic-ci-openshift-release-main-ci-4.22-e2e-aws-ovn-techpreview-serial-1of3: dbd9ca29-0561-463d-811c-3e34e78a5978
{  1 events happened too frequently

event happened 22 times, something is wrong: namespace/openshift-config-operator node/ip-10-0-111-205.ec2.internal pod/openshift-config-operator-7cdbb85bd7-5n8md hmsg/440100f6ee - reason/ProbeError Liveness probe error: Get "https://10.129.0.30:8443/healthz": dial tcp 10.129.0.30:8443: connect: connection refused result=reject 
body: 
 (20:15:44Z)}


https://prow.ci.openshift.org/view/gs/test-platform-results/logs/openshift-cluster-network-operator-2944-ci-4.22-upgrade-from-stable-4.21-e2e-aws-ovn-upgrade/2050257487575126016
The phase of Pod resize-test-22b4h is Failed which is unexpected.

Don't see anything that can be somehow influenced by this PR

**Comment by tsorya** [comment]
periodic-ci-openshift-release-main-ci-4.22-e2e-aws-upgrade-ovn-single-node:
1. run failed on `"kube-apiserver" in namespace: "openshift-kube-apiserver" for revision: 6 on node: "ip-10-0-115-216.us-west-2.compute.internal" didn't show up` though this is test issue as revision 7 was there.
2. 3 tests failed on [sig-node] Pod InPlace Resize Container guaranteed qos - resizing mem
3. others failed on `event happened 27 times, something is wrong: namespace/openshift-machine-config-operator /machine-config hmsg/cd9e602a3c - reason/OperatorDegraded: MachineConfigNodeFailed Failed to resync 4.22.0-0.ci-2026-05-02-032308 because: Get "https://172.30.0.1:443/apis/machineconfiguration.openshift.io/v1/machineconfignodes": dial tcp 172.30.0.1:443: connect: connection refused (19:29:09Z) result=reject` or similar, which can happen on SNO upgrade cluster when etcd or kubeapi is not reachable and seems this this SNO vm was pretty slow


periodic-ci-openshift-release-main-ci-4.22-upgrade-from-stable-4.21-e2e-gcp-ovn-rt-upgrade

1. 6/9 runs (1, 2, 4, 5, 8, 9) fail purely on the no-default-service-account monitor test -- nothing else
2. 3/9 runs (3, 7, 10) also have 1 additional blocking e2e conformance test failure each (all different tests -- sig-cli, sig-api-machinery -- none networking-related


periodics-e2e-aws-ovn-conformance

1. 4 tests failed on [sig-etcd] etcd leader changes are not excessive
2. 2 tests failed on 
[sig-node] [DRA] kubelet DynamicResourceAllocation - inline claim referenced by multiple containers
